src/dbnode/storage/namespace.go

```diff
@@ -938,20 +940,112 @@ func (n *dbNamespace) Flush(
-		if err := shard.Flush(blockStart, flushPersist, nsCtx); err != nil {
+		if err := shard.WarmFlush(blockStart, flushPersist, nsCtx); err != nil {
 			detailedErr := fmt.Errorf("shard %d failed to flush data: %v",
 				shard.ID(), err)
 			multiErr = multiErr.Add(detailedErr)
 		}
 	}
 
 	res := multiErr.FinalError()
-	n.metrics.flush.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	n.metrics.flushWarmData.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	return res
+}
+
+// idAndBlockStart is the composite key for the genny map used to keep track of
+// dirty series that need to be ColdFlushed.
+type idAndBlockStart struct {
+	id         ident.ID
+	blockStart xtime.UnixNano
+}
+
+type coldFlushReuseableResources struct {
+	// dirtySeries is a map from a composite key of <series ID, block start>
+	// to an element in a list in the dirtySeriesToWrite map. This map is used
+	// to quickly test whether a series is dirty for a particular block start.
+	//
+	// The composite key is deliberately made so that this map stays one level
+	// deep, making it easier to share between shard loops, minimizing the need
+	// for allocations.
+	//
+	// Having a reference to the element in the dirtySeriesToWrite list enables
+	// efficient removal of the series that have been read and subsequent
+	// iterating through remaining series to be read.
+	dirtySeries *dirtySeriesMap
+	// dirtySeriesToWrite is a map from block start to a list of dirty series
+	// that have yet to be written to disk.
+	dirtySeriesToWrite map[xtime.UnixNano]*idList
+	// idElementPool is a pool of list elements to be used when constructing
+	// new lists for the dirtySeriesToWrite map.
+	idElementPool *idElementPool
+	fsReader      fs.DataFileSetReader
+}
+
+func (r *coldFlushReuseableResources) reset() {
+	for _, seriesList := range r.dirtySeriesToWrite {
+		if seriesList != nil {
+			seriesList.Reset()
+		}
+		// Don't delete the empty list from the map so that other shards don't
+		// need to reinitialize the list for these blocks.
+	}
+
+	r.dirtySeries.Reset()
+}
+
+func (n *dbNamespace) ColdFlush(
+	flushPersist persist.FlushPreparer,
+	latestWarmFlushTime time.Time,
+) error {
+	// NB(rartoul): This value can be used for emitting metrics, but should not be used
+	// for business logic.
+	callStart := n.nowFn()
+
+	n.RLock()
+	if n.bootstrapState != Bootstrapped {
+		n.RUnlock()
+		n.metrics.flushColdData.ReportError(n.nowFn().Sub(callStart))
+		return errNamespaceNotBootstrapped
+	}
+	nsCtx := namespace.Context{Schema: n.schemaDescr}
+	n.RUnlock()
+
+	if !n.nopts.ColdWritesEnabled() {
+		n.metrics.flushColdData.ReportSuccess(n.nowFn().Sub(callStart))
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.GetOwnedShards()
+
+	fsReader, err := fs.NewReader(n.opts.BytesPool(), n.opts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return err
+	}
+	resources := coldFlushReuseableResources{
```

Comment:
Can you just make a constructor function for this? Its a little tedious to be reading through the logic for `ColdFlushes` and be wondering why you're constructing an `fsReader`